### PR TITLE
Answer detection

### PR DESCRIPTION
**Pull request by denisnikulin**
answer detection - emitting call.bridge event

### CODE REVIEW

File: src/callServer.js [hunk 1]
```diff
@@ -11,6 +11,11 @@ class CallServer extends EventEmitter {
     this._driver.on('call.ended', (callId) => {
       this.emit('call.ended', callId);
     })
+
+    this._driver.on('call.bridged', (callId) => {
+      this.emit('call.bridged', callId);
+    })
```
Comment (basia-horwath) [inline src/callServer.js]: 👍

File: src/drivers/esl.js [hunk 2]
```diff
@@ -20,6 +21,10 @@ class EslDriver extends CallDriver {
         emitEnd(callId);
       });
 
+      socket.on('CHANNEL_ANSWER', (...args) => {
+        emitBridged(callId);
+      });
```
Comment (basia-horwath) [inline src/drivers/esl.js]: 👍

File: src/drivers/mockSocket.js [hunk 3]
```diff
@@ -16,6 +16,10 @@ class MockSocketDriver extends CallDriver {
         this.emit('call.ended', callId);
       })
 
+      socket.on('call.bridged', () => {
+        this.emit('call.bridged', callId);
+      })
```
Comment (basia-horwath) [inline src/drivers/mockSocket.js]: 👍

File: src/servers/esl.js [hunk 4]
```diff
@@ -17,11 +17,15 @@ class EslServer extends SocketServer {
...
+    this._getSocket(callId).filter('variable_conference', callId);
+    return this._getSocket(callId).api('conference ' + callId + '++flags{moderator} dial {conference=' + callId + '}sofia/gateway/partner/' + address)
   }
 
   terminateCall(callId) {
-    return this._getSocket(callId).api('conference ' + callId + ' hup all')
+    this._getSocket(callId).event_json('CHANNEL_HANGUP');
+    this._getSocket(callId).filter('variable_conference_name', callId);
+    return this._getSocket(callId).execute('conference ' + callId + ' hup all')
   }
```
Comment (basia-horwath) [inline src/servers/esl.js]: So, are both _bridgeCall()_ and _ terminateCall()_ working now? :)

One question here: you're using _ variable_conference_ in the filter for bridging and _ variable_conference_name_ in the filter for termination. Is that correct?

Comment (denisnikulin) [inline src/servers/esl.js]: yes, freeswitch prefixes variable_ itself. 
all works. I will need to test cancelling the call before bridge though.

Comment (basia-horwath) [inline src/servers/esl.js]: Ok, but then, why is the variable _conference_name_ in the termination filter, while it is _conference_ everywhere else? ;)

File: src/servers/esl.js [hunk 5]
```diff
@@ -13,15 +13,19 @@ class EslServer extends SocketServer {
...
 
   holdCall(callId) {
+    this._getSocket(callId).event_json('CHANNEL_HANGUP');
+    this._getSocket(callId).filter('variable_conference_name', callId);
```
Comment (basia-horwath) [inline src/servers/esl.js]: Now, again ;) , why is the variable _conference_name_ here, while it is _conference_ everywhere else? 😉

Comment (denisnikulin) [inline src/servers/esl.js]: https://github.com/utilitywarehouse/uw-lib-callcontrol/pull/2#discussion_r133983969

freeswitch prefixes it automatically!:)

Comment (denisnikulin) [inline src/servers/esl.js]: https://circleci.com/gh/utilitywarehouse/uw-lib-callcontrol

File: src/servers/mockSocket.js [hunk 6]
```diff
@@ -14,7 +14,8 @@ class MockSocketServer extends SocketServer {
 
   bridgeCall(callId, address) {
     const socket = this._getSocket(callId);
-    return Promise.resolve(socket.write(`Bridging ${callId} to ${address}\n`, "utf8"));
+    return Promise.resolve(socket.write(`Bridging ${callId} to ${address}\n`, "utf8"))
+      .then(socket.emit('call.bridged'));
```
Comment (basia-horwath) [inline src/servers/mockSocket.js]: This works, but looks weird. We don't need to do _then_ here, I think, 'cause we are not dependent on the result of the previous operation.

Actually, we could probably get rid of the whole Promise entirely. Try:
```
bridgeCall(callId, address) {
    const socket = this._getSocket(callId);
    socket.write(`Bridging ${callId} to ${address}\n`, "utf8");
    socket.emit('call.bridged');
}
```
If it works, get rid of the Promises in other methods as well :)

File: src/servers/esl.js (hunk 5 above)
Comment (basia-horwath) [inline src/servers/esl.js]: I understand that FS prefixes with _variable__. My point is: your variable is _conference_name_ here, but it is _conference_ in the next line:
```
return this._getSocket(callId).execute('conference', callId + '@video-mcu-stereo++flags{mintwo}')
```
and in the _bridgeCall()_ method.

This seems suspicious to me, though I might not be getting sth here :)

Comment (basia-horwath) [inline src/servers/esl.js]: Here:
```
this._getSocket(callId).filter('variable_conference_name', callId);
```
vs there:
```
this._getSocket(callId).filter('variable_conference', callId);
```

File: tests/callServer.spec.js [hunk 7]
```diff
@@ -62,4 +62,12 @@ describe('CallServer', () => {
     driver.emit('call.started', ({callId, socket}));
   })
 
+  it('emits call.bridge event when the call has been bridged!', (done) => {
+    const callId = 1;
+    const socket = {};
+
+    callServer.on('call.bridged', done);
+    driver.emit('call.bridged', ({callId, socket}));
```
Comment (basia-horwath) [inline tests/callServer.spec.js]: This works ('cause the test does not verify the arguments emitted by the driver), but technically, the driver should only emit callId as the argument. We know that from here: https://github.com/utilitywarehouse/uw-lib-callcontrol/blob/answer-detection/src/callServer.js#L15. 
So:
```
driver.emit('call.bridged', callId);
```

File: src/servers/esl.js [hunk 8]
```diff
@@ -13,15 +13,18 @@ class EslServer extends SocketServer {
...
-    return this._getSocket(callId).api('conference ' + callId + '++flags{moderator} dial sofia/gateway/partner/' + address)
+    this._getSocket(callId).event_json('CHANNEL_ANSWER');
+    this._getSocket(callId).filter('variable_conference', callId);
+    return this._getSocket(callId).api('conference ' + callId + '++flags{moderator} dial {conference=' + callId + '}sofia/gateway/partner/' + address)
   }
 
   terminateCall(callId) {
-    return this._getSocket(callId).api('conference ' + callId + ' hup all')
+    return this._getSocket(callId).execute('conference ' + callId + ' hup all')
   }
```
Comment (basia-horwath) [inline src/servers/esl.js]: 👍

**Pull request merged**